Docs for API Platform 2.6 (#1255)
api-platform/docs: distribution/index.md

# Getting Started With API Platform: Create Your API and Your Jamstack Site!

![The welcome page](images/api-platform-2.5-welcome.png)

> *API Platform* is the most advanced API platform, in any framework or language.
>
> —Fabien Potencier (creator of Symfony), SymfonyCon 2017

[API Platform](https://api-platform.com) is a powerful yet easy to use **full stack** framework dedicated to API-driven projects and implementing the [Jamstack](https://jamstack.org/) architecture.

## Introduction

API Platform contains [a **PHP** library (Core)](../core/index.md) to create fully featured hypermedia (or [GraphQL](../core/graphql.md)) web APIs supporting industry-leading standards: [JSON-LD](https://json-ld.org/) with [Hydra](https://www.hydra-cg.com/), [OpenAPI](../core/swagger.md)...

> It also provides ambitious **JavaScript** tools to create web and mobile apps based on the most popular frontend technologies in a snap. To do so, these tools parse the documentation of the API (or of any other API supporting Hydra or OpenAPI).

Reviewer: ```suggestion
It provides ambitious **JavaScript** tools to create web and mobile applications based on the most popular frontend technologies in a snap. These tools parse the documentation API's or any other API supporting Hydra or OpenAPI.
```

Author: API's ?

 API Platform is also shipped with **[Docker](../deployment/docker-compose.md)** and **[Kubernetes](../deployment/kubernetes.md)** definitions, to develop and deploy instantly on the cloud.

The easiest and most powerful way to get started is [to download the API Platform distribution](https://github.com/api-platform/api-platform/releases). It contains:

> * the API skeleton, including with [the Core library](../core/index.md), [the Symfony framework](https://symfony.com/) ([optional](../core/bootstrap.md)) and [the Doctrine ORM](https://www.doctrine-project.org/projects/orm.html) ([optional](../core/extending.md))

Reviewer 2: ```suggestion
* the API skeleton, including [the Core library](../core/index.md), [the Symfony framework](https://symfony.com/) ([optional](../core/bootstrap.md)) and [the Doctrine ORM](https://www.doctrine-project.org/projects/orm.html) ([optional](../core/extending.md))
```

* [the client scaffolding tool](../client-generator/) to generate [Next.js](../client-generator/
) web applications from the API documentation ([Nuxt.js](https://nuxtjs.org/), [Vue](https://vuejs.org/), [Create React App](https://reactjs.org), [React Native](https://facebook.github.io/react-native/), [Quasar](https://quasar.dev/) and [Vuetify](https://vuetifyjs.com/) are also supported)
* [a beautiful admin interface](../admin/), built on top of React Admin, dynamically created by parsing the API documentation
* all you need to [create real-time and async APIs using the Mercure protocol](../core/mercure.md)
* a [Docker](../deployment/docker-compose.md) definition to start a working development environment in a single command, providing containers for the API and the Next.js web application
* a [Helm](https://helm.sh/) chart to deploy the API in any [Kubernetes](../deployment/kubernetes.md) cluster

## A Bookshop API

To discover how the framework works, we will create an API to manage a bookshop.

> To create a fully featured API, an admin interface and a Progressive Web App using Next.js, we will only have to design **the public data

Reviewer: ```suggestion
To create a fully featured API, an admin interface and a Progressive Web App using Next.js, all you need is to design **the public data
```

`https://localhost:8443/greetings.json` to retrieve data in raw JSON.

Of course, you can also use your favorite HTTP client to query the API.

> We are fond of [Postman](https://hoppscotch.io/). It works perfectly well with API Platform, has native OpenAPI support,

Reviewer: hoppscotch ?

Author: oops

Books have an id, an ISBN, a title, a description, an author, a publication date and are related to a list of reviews.
Reviews have an id, a rating (between 0 and 5), a body, an author, a publication date and are related to one book.

Let's describe this data model as a set of Plain Old PHP Objects (POPO):

```php
<?php
// api/src/Entity/Book.php

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;
use Doctrine\Common\Collections\ArrayCollection;

/** A book. */
#[ApiResource]
class Book
{
    /** The id of this book. */

>     private ?int $id = null;

Reviewer: why is the id nullable?

Author: same

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;

/** A review of a book. */
#[ApiResource]
class Review
{
    /** The id of this review. */

>     private ?int $id = null;

Reviewer: why is the id nullable?

Author: To prevent issues if this property is assessed before an ID is generated by the DBMS.

}
```

> We created two typical PHP objects with the corresponding PHPDoc, both marked with the `#[ApiResource]` attribute. For convenience, we also used the Doctrine Collection library (that is independent from Doctrine ORM), but it's not mandatory.

Reviewer 2: ```suggestion
We created two typical PHP objects, both marked with the `#[ApiResource]` attribute. For convenience, we also used the Doctrine Collection library (that is independent from Doctrine ORM), but it's not mandatory.
```

Author: It's important to say that we added the PHPDoc isn't it?

Reviewer 2: But where did we do that?

Author: On every properties. Ex: `/** The date of publication of this review.*/`

Reload `https://localhost/docs/`: API Platform used these classes to generate an OpenAPI documentation (a Hydra documentation is also exposed), and registered for us [the typical REST routes](../core/operations.md).

Note that entities' and properties' descriptions
included in their PHPDoc appear in the API documentation, and that API Platform use PHP types to generate the appropriate JSON Schemas.

The framework also use these metadata to serialize and deserialize data from JSON (and other formats) to PHP objects (back and forth)!

For the sake of simplicity, in this example we used public properties (except for the id, see below). API Platform (as well
as Symfony and Doctrine) also supports accessor methods (getters/setters), use them if you want to.

> We used a private property and a getter for the id to enforce the fact that it is read only (we will let the DBMS generating it). API Platform also has first-grade support for UUIDs. [You should

Reviewer: maybe offer to install `symfony/uid`?

Author: Maybe not in this guide (it's already a bit bloated), we need to add a new file for this.

probably use them instead of auto-incremented ids](https://www.clever-cloud.com/blog/engineering/2015/05/20/why-auto-increment-is-a-terrible-idea/).

Because API Platform provides all the infrastructure for us, our API is almost ready! The only remaining task to have a working API is to be able to query and persist data.

## Plugging the Persistence System

To retrieve and save data, API Platform proposes two main options (and we can mix them):

> 1. Writing our own [data providers](../core/data-providers.md) and [data persisters](../core/data-persisters.md) to fetch and save data in any persistence system and trigger our custom business logic. This is what we recommend if you want to separate the public data model exposed by the API from the internal one, and to implement a layered architecture such as Clean Architecture or Hexagonal Architecture ;

Reviewer 2: ```suggestion
1. Writing our own [data providers](../core/data-providers.md) and [data persisters](../core/data-persisters.md) to fetch and save data in any persistence system and trigger our custom business logic. This is what we recommend if you want to separate the public data model exposed by the API from the internal one, and to implement a layered architecture such as Clean Architecture or Hexagonal Architecture;
```

2. Using one of the various existing data providers and persisters allowing to automatically fetch and persist data using popular persistence libraries. Out of the box, data providers and persisters are provided for Doctrine ORM] and [Doctrine MongoDB ODM](../core/mongodb.md). A data provider (but no persister yet) is also available for [Elasticsearch](../core/elasticsearch.md). [Pomm](https://github.com/pomm-project/pomm-api-platform) and [PHP Extended SQL](https://github.com/soyuka/esql#api-platform-bridge) also provides data providers and persisters for API Platform. We recommend this approach for Rapid Application Development.

Be sure to read the [General Design Considerations](../core/design.md) document to learn more about the architecture of API Platform and how to choose between these two approaches.

> For the sake of simplicity, we will use the built-in Doctrine ORM data provider in the rest of this tutorial.

Reviewer: ```suggestion
Here, we will use the built-in Doctrine ORM data provider in the rest of this tutorial.
```

+     */
    public ?\DateTimeInterface $publicationDate = null;

-    /** The book this review is about. */
+    /**
+     * The book this review is about.
+     *
+     * @ORM\ManyToOne(targetEntity="Book", inversedBy="reviews")
+     */
    public Book $book;
```

**Tip**: you can also use Symfony [MakerBundle](https://symfonycasts.com/screencast/symfony-fundamentals/maker-command?cid=apip) thanks to the `--api-resource` option:

    $ docker-compose exec php bin/console make:entity --api-resource

Doctrine's annotations map these entities to tables in the database. Annotations are convenient as they allow grouping
the code and the configuration but, if you want to decouple classes from their metadata, you can switch to XML or YAML mappings.
They are supported as well.

**Note:** Doctrine ORM doesn't support PHP 8 attributes at this time, but support should be added soon ([a Pull Request already exists](https://github.com/doctrine/orm/pull/8266)).

Learn more about how to map entities with the Doctrine ORM in [the project's official documentation](https://docs.doctrine-project.org/projects/doctrine-orm/en/latest/reference/association-mapping.html)

> or in Kévin's book "[Persistence in PHP with the Doctrine ORM](https://www.amazon.fr/gp/product/B00HEGSKYQ/ref=as_li_tl?ie=UTF8&camp=1642&creative=6746&creativeASIN=B00HEGSKYQ&linkCode=as2&tag=kevidung-21)".

Reviewer: ```suggestion
or in Kévin Dunglas's book "[Persistence in PHP with the Doctrine ORM](https://www.amazon.fr/gp/product/B00HEGSKYQ/ref=as_li_tl?ie=UTF8&camp=1642&creative=6746&creativeASIN=B00HEGSKYQ&linkCode=as2&tag=kevidung-21)".
```

API Platform also has an awesome [client generator](../client-generator/index.md) able to scaffold fully working Next.js, Nuxt.js, React/Redux, Vue.js, Quasar and Vuetify Progressive Web Apps that you can easily tune and customize. The generator also supports
[React Native](https://facebook.github.io/react-native/) if you prefer to leverage all capabilities of mobile devices.

The distribution comes with a skeleton ready to welcome the [Next.js](https://nextjs.org/) flavor of the generated code. To bootstrap your app, run:

    $ docker-compose exec pwa generate-api-platform-client

Open `https://localhost/greetings/` in your browser:

![The Next.js Progressive Web App](images/api-platform-2.5-pwa-react.png)

You can also choose to generate the code for a specific resource with the `--resource` argument (example:
`generate-api-platform-client --resource books`).

The generated code contains a list (including pagination), a delete button, a creation and an edition form. It also includes
[Bootstrap](https://getbootstrap.com) markup and [ARIA roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA)
to make the app usable by people with disabilities.

> If you prefer to generate a PWA built on top of an other frontend stack, read [the dedicated documentation](../client-generator/index.md).

Reviewer 2: ```suggestion
If you prefer to generate a PWA built on top of another frontend stack, read [the dedicated documentation](../client-generator/index.md).
```
